coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go

```diff
@@ -0,0 +1,282 @@
+	}
+
+	options := &redis.Options{
+		Addr:            r.Config.Host,
+		Password:        r.Config.Password,
+		DB:              0,
+		MaxRetries:      3,
+		MaxRetryBackoff: time.Second * 2,
+	}
+	if r.Config.RequiresTLS {
+		options.TLSConfig = &tls.Config{
+			InsecureSkipVerify: !r.Config.RequiresTLS,
+		}
+	}
+	client := redis.NewClient(options)
+	if _, err := client.Ping().Result(); err != nil {
+		rLog.Debugf("  P (Redis State): failed to connect to redis %+v", err)
+		return v1alpha2.NewCOAError(err, fmt.Sprintf("redis stream: error connecting to redis at %s", r.Config.Host), v1alpha2.InternalError)
+	}
+	r.Client = client
+	rLog.Debug("  P (Redis State): Successfully launch redis state provider")
+	return nil
+}
+
+func (r *RedisStateProvider) Upsert(ctx context.Context, entry states.UpsertRequest) (string, error) {
+	_, span := observability.StartSpan("Redis State Provider", ctx, &map[string]string{
+		"method": "Upsert",
+	})
+
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	keyPrefix := getKeyNamePrefix(entry.Metadata)
+	rLog.Debugf("  P (Redis State): upsert states %s with keyPrefix %s, traceId: %s", entry.Value.ID, keyPrefix, span.SpanContext().TraceID().String())
+
+	key := fmt.Sprintf("%s%s%s", keyPrefix, separator, entry.Value.ID)
+	var body []byte
+	body, err = json.Marshal(entry.Value.Body)
+	if err != nil {
+		return entry.Value.ID, err
+	}
+	properties := map[string]interface{}{
+		"values": string(body),
+		"etag":   entry.Value.ETag,
+	}
+	_, err = r.Client.HSet(key, properties).Result()
+	return entry.Value.ID, err
+}
+
+func (r *RedisStateProvider) List(ctx context.Context, request states.ListRequest) ([]states.StateEntry, string, error) {
+	_, span := observability.StartSpan("Redis State Provider", ctx, &map[string]string{
+		"method": "List",
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	var entities []states.StateEntry
+	keyPrefix := getKeyNamePrefix(request.Metadata)
+	rLog.Debugf("  P (Redis State): list states with keyPrefix %s, traceId: %s", keyPrefix, span.SpanContext().TraceID().String())
+
+	filter := fmt.Sprintf("%s%s*", keyPrefix, separator)
+	var cursor uint64 = 0
+	var keys []string
+
+	for {
+		var err error
+		keys, cursor, err = r.Client.Scan(cursor, filter, entryCountPerList).Result()
+		if err != nil {
+			rLog.Errorf("  P (Redis State): failed to get all the keys matching pattern %s: %+v", keyPrefix, err)
+		}
+
+		for _, key := range keys {
+			result, err := r.Client.HGetAll(key).Result()
+			if err != nil || len(result) == 0 {
+				rLog.Errorf("  P (Redis State): failed to get entry for key %s: %+v", key, err)
+				continue
+			}
+			parts := strings.Split(key, separator)
+			if len(parts) != 3 {
+				rLog.Errorf("  P (Redis State): key is not valid %s: %+v", key, err)
+				continue
+			}
+			entry, err := CastRedisPropertiesToStateEntry(parts[2], result)
```

Comment:
We need to support filters in list quest. Please see memory state provider or k8s state provider as an example

Reply:
Thanks Haishi for bringing this up. I have looked into secondary index for Redis previously and it's supported by RediSearch which is published by Redis Lab and uses Redis Source Available license. So it's not proper to use the product in Symphony. If we want to have the filters, I am afraid we need to list all the items into memory and do filter by symphony. Do you think it's acceptable or not?

Reply:
Behavior consistency is important as other parts of the system would expect filters to work when Redis is used. We can worry about performance at a later time.

Reply:
Add filtering in List interface using shared in-memory filter as memory state provider. Also respect UpdateStateOnly field in Upsert interface.